### PR TITLE
feat(data): add support for FP8 target cache quantization (float8_e4m3fn & float8_e5m2)

### LINKED ISSUE #26: feat: add support for FP8 target cache quantization
# Feature Description
Currently, preparing the target cache (saving the intermediate target hidden states and last hidden states) requires a significant disk footprint. For large-scale pretraining or instruction-tuning datasets, target caches can easily scale up to multiple terabytes of storage because activations are stored in 16-bit formats (`bfloat16`). 

I would like to propose adding support for **FP8 quantization** (`float8_e4m3fn` and `float8_e5m2`) to reduce the target cache storage footprint by **50%** (1 byte per element instead of 2 bytes).

---

## Proposed Solution
The implementation will introduce:
1. **Configurable CLI parameter**: Add a `--hidden-dtype` argument to `prepare_target_cache.py` to allow users to generate caches in `bfloat16`, `float8_e4m3fn`, or `float8_e5m2`.
2. **Quantized Serialization**: Quantize the hidden state tensors to the specified FP8 format and save them as raw bytes in the cache shards.
3. **Transparent Dequantization**: Read the float8 bytes back as float8 tensors and cast them back to `bfloat16` on the fly inside `CacheDataset.__getitem__`. This prevents any downstream model or trainer code from having to be modified, as they will still receive standard `bfloat16` tensors.
4. **Full Backward Compatibility**: Caches generated in older versions (or without specifying a dtype) will default to `bfloat16` and load normally.

---

## Implementation Details & Impact
I have already implemented a working prototype of this and validated it:
* **Storage savings**: Exactly 50% reduction in shard file sizes on disk.
* **Precision impact**: 
  * `float8_e4m3fn` introduces a Mean Absolute Error (MAE) of only `~0.0090` (near-lossless).
  * `float8_e5m2` introduces an MAE of only `~0.0179`.
* **Testing**: Includes a comprehensive suite of unit tests covering precision degradation, dataloader collation, and end-to-end integration tests.

I'd love to get feedback on this feature and submit a pull request!

### PR DESCRIPTION
**Pull request by ApurveKaranwal**
# feat(data): add support for FP8 target cache quantization (float8_e4m3fn & float8_e5m2)

## Description / Issue Reference
* **Closes #26 

## Summary
This PR implements quantization support (`float8_e4m3fn` and `float8_e5m2`) for the target cache preparation and dataloading pipelines. 

By saving the target model's intermediate hidden states and last hidden states in FP8 instead of BF16, we achieve a **50% reduction in disk storage requirements** (1 byte per element instead of 2 bytes). Dequantization to `bfloat16` happens transparently during dataset loading (`__getitem__`), meaning the downstream training collator (`CacheCollator`), trainers, and loss functions require **zero modifications** to run.

---

### 1. Quantization & Transparent Dequantization
* **File**: `deepspec/data/target_cache_dataset.py`
  * Extended `validate_target_cache_manifest` to support `float8_e4m3fn` and `float8_e5m2` as valid `hidden_dtype` values.
  * Added `_tensor_to_float8_bytes` using raw `uint8` byte views of PyTorch float8 tensors (`view(torch.uint8)`) for safe writing, bypassing NumPy's lack of native float8 dtypes.
  * Updated `expected_target_cache_tensor_nbytes` to dynamically scale the expected size based on the chosen dtype (1 byte per element for FP8, 2 bytes for BF16).
  * Added `_read_float8_tensor_from_shard` to read quantized raw bytes, cast them back into the original float8 tensor view, and transparently dequantize them to `bfloat16` on demand during `__getitem__`.

### 2. CLI Configurable Cache Dtype
* **File**: `scripts/data/prepare_target_cache.py`
  * Added the `--hidden-dtype` CLI argument (choices: `bfloat16`, `float8_e4m3fn`, `float8_e5m2`, defaulting to `bfloat16`).
  * Threaded the selected dtype through the async writers and metadata manifest builders.

### 3. Tests Package
* **Files**: `tests/__init__.py`, `tests/test_quantization.py`
  * Created a comprehensive suite containing 5 distinct test classes verifying:
    * Baseline correctness (`bfloat16` behaves identically to the original input).
    * `float8` file size reduction (verifying a exact 50% storage savings on disk).
    * Precision metrics (validating numerical closeness to original activations).
    * Batch collation and padding compatibility in the `CacheCollator`.
    * End-to-end `prepare_target_cache.py` run on CPU with a mock distributed setup, model, and tokenizer.

---

## Backward Compatibility
* Existing target caches are fully compatible. The loading pipeline fetches `hidden_dtype` from the JSON manifest and defaults to `bfloat16` if the field is missing.

---

## Verification and Test Results

All unit tests pass successfully. The numerical delta (Mean Absolute Error) introduced by quantization is minimal, ensuring downstream training stability:

```bash
Ran 5 tests in 36.983s

OK
FP8 E4M3FN target_hidden_states MAE: 0.009033203125
FP8 E4M3FN target_last_hidden_states MAE: 0.009033203125
FP8 E5M2 target_hidden_states MAE: 0.0179443359375
FP8 E5M2 target_last_hidden_states MAE: 0.0179443359375

### CODE REVIEW

**Comment by ApurveKaranwal** [comment]
@DaoyuanLi2816 sir, please review my PR and let me know if any changes are required.
